logbook-core/src/main/java/org/zalando/logbook/HeaderFilters.java

```diff
@@ -28,52 +31,78 @@ public static HeaderFilter defaultValue() {
+
+        return headers -> headers.apply(names, (name, previous) ->
+                previous == null ?
+                        null :
+                        nCopies(previous.size(), replacement));
     }
 
-    public static HeaderFilter eachHeader(final BinaryOperator<String> operator) {
-        return headers -> {
-            final Map<String, List<String>> result = Headers.empty();
+    public static HeaderFilter replaceHeaders(
+            final Predicate<String> keyPredicate,
+            final String replacement) {
+
+        return replaceHeaders(predicate(keyPredicate), replacement);
+    }
 
-            headers.forEach((name, values) ->
-                    result.put(name, values.stream()
-                            .map(value -> operator.apply(name, value))
-                            .collect(toList())));
+    public static HeaderFilter replaceHeaders(
+            final BiPredicate<String, String> predicate,
+            final String replacement) {
 
-            return Headers.immutableCopy(result);
-        };
+        return eachHeader((key, value) ->
+                predicate.test(key, value) ? replacement : value);
     }
 
-    public static HeaderFilter removeHeaders(final Predicate<String> keyPredicate) {
-        return removeHeaders((key, value) -> keyPredicate.test(key));
+    public static HeaderFilter eachHeader(
+            final BinaryOperator<String> operator) {
+
+        return headers -> headers.apply(headers.keySet(), (name, values) ->
+                values.stream()
+                        .map(value -> operator.apply(name, value))
+                        .collect(toList()));
     }
 
-    public static HeaderFilter removeHeaders(final BiPredicate<String, String> predicate) {
-        return headers -> {
-            final Map<String, List<String>> result = Headers.empty();
+    public static HeaderFilter removeHeaders(
+            final String... names) {
 
-            headers.forEach((name, original) -> {
-                final List<String> values = original.stream()
-                        .filter(value -> !predicate.test(name, value))
-                        .collect(toList());
+        return headers -> headers.delete(names);
+    }
 
-                if (values.isEmpty()) {
-                    return;
-                }
+    public static HeaderFilter removeHeaders(
+            final Predicate<String> predicate) {
 
-                result.put(name, values);
-            });
+        return headers -> headers.delete(predicate(predicate));
+    }
 
-            return Headers.immutableCopy(result);
-        };
+    public static HeaderFilter removeHeaders(
+            final BiPredicate<String, String> predicate) {
+
+        return headers -> headers.apply((name, previous) -> {
+            final List<String> next = previous.stream()
+                    .filter(value -> predicate.negate().test(name, value))
+                    .collect(toList());
+
+            if (next.isEmpty()) {
```

Comment:
What if it was empty before and our predicate didn't actually match anything? Should we keep the empty list in that case?

Reply:
(i.e., should we swap the first two ifs?